sdks/go/pkg/beam/x/hooks/perf/perf.go

```diff
@@ -0,0 +1,105 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package perf
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"runtime/pprof"
+
+	"github.com/apache/beam/sdks/go/pkg/beam/core/runtime"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/runtime/harness"
+	fnpb "github.com/apache/beam/sdks/go/pkg/beam/model/fnexecution_v1"
+)
+
+var (
+	sessionCaptureHookKey = harness.HookKey("session_capture")
+	traceCaptureHookKey   = harness.HookKey("trace_capture")
+)
+
+// TraceCaptureHook is used by the harness to have the runner
+// persist a trace record with the supplied name and comment.
+// The type of trace can be determined by the prfix of the string
+// cpu: A profile compatible with traces produced by runtime/pprof
+// trace: A trace compatible with traces produced by runtime/trace
+// TODO(wcn): define and document future formats here.
+type TraceCaptureHook func(string, io.Reader) error
+
+var traceCaptureHookRegistry = make(map[string]TraceCaptureHook)
+
+// RegisterTraceCaptureHook registers a TraceCaptureHook for the
+// supplied identifier. It panics if the same identifier is
+// registered twice.
+func RegisterTraceCaptureHook(name string, c TraceCaptureHook) {
+	if _, exists := traceCaptureHookRegistry[name]; exists {
+		panic(fmt.Sprintf("RegisterTraceCaptureHook: %s registered twice", name))
+	}
+	traceCaptureHookRegistry[name] = c
+
+	if len(traceCaptureHookRegistry) == 1 {
+		var cpuProfBuf bytes.Buffer
+		harness.RegisterRequestHook("trace", func(_ context.Context, _ *fnpb.InstructionRequest) error {
+			cpuProfBuf.Reset()
+			return pprof.StartCPUProfile(&cpuProfBuf)
```

Comment:
<!--new_thread; commit:d3b9b8992429b98ae5429208961299a90dcdec50; resolved:0-->
We're using Trace/"trace" all over this code, but it's measuring CPU profiles, not traces. Is this copy pasta? Should there be another set of CPU methods that are presently missing from this PR?

Reply:
Yes, I needed to fix that. This was a point-in-time snapshot. There are still things to address.

My next push will fix open issues and have better markers for the WIP stuff, but the next push won't be the "final" version either.